src/hypothesis/searchstrategy/regex.py

```diff
@@ -0,0 +1,378 @@
+
+        elif code == sre.IN:
+            # Regex '[abc0-9]' (set of characters)
+            charsets = value
+
+            builder = CharactersBuilder(negate=charsets[0][0] == sre.NEGATE,
+                                        flags=context.flags)
+
+            for charset_code, charset_value in charsets:
+                if charset_code == sre.NEGATE:
+                    # Regex '[^...]' (negation)
+                    pass
+                elif charset_code == sre.LITERAL:
+                    # Regex '[a]' (single char)
+                    builder.add_chars(hunichr(charset_value))
+                elif charset_code == sre.RANGE:
+                    # Regex '[a-z]' (char range)
+                    low, high = charset_value
+                    for char_code in hrange(low, high + 1):
+                        builder.add_chars(hunichr(char_code))
+                elif charset_code == sre.CATEGORY:
+                    # Regex '[\w]' (char category)
+                    builder.add_category(charset_value)
+                else:  # pragma: no cover
+                    # Currently there are no known code points other than
+                    # handled here. This code is just future proofing
+                    raise InvalidArgument(
+                        'Unknown charset code: %s' % charset_code
+                    )
+
+            if not is_text:
+                return builder.strategy.map(lambda u: u.encode('utf-8'))
+            return builder.strategy
+
+        elif code == sre.ANY:
+            # Regex '.' (any char)
+            if is_text:
+                if context.flags & re.DOTALL:
+                    return hs.characters()
+                return hs.characters(blacklist_characters=u'\n')
+            else:
+                if context.flags & re.DOTALL:
+                    return binary_char
+                return binary_char.filter(lambda c: c != b'\n')
+
+        elif code == sre.AT:
+            # Regexes like '^...', '...$', '\bfoo', '\Bfoo'
+            # An empty string (or newline) will match the token itself, but
+            # we don't and can't check the position (eg '%' at the end)
+            if value == sre.AT_END:
+                return hs.one_of(hs.just(EMPTY),
+                                 hs.just(u'\n' if is_text else b'\n'))
+            return hs.just(EMPTY)
+
+        elif code == sre.SUBPATTERN:
+            # Various groups: '(...)', '(:...)' or '(?P<name>...)'
+            old_flags = context.flags
+            if HAS_SUBPATTERN_FLAGS:  # pragma: no cover
+                # This feature is available only in specific Python versions
+                context.flags = (context.flags | value[1]) & ~value[2]
+
+            strat = _strategy(value[-1], context, is_text)
+
+            context.flags = old_flags
+
+            if value[0]:
+                context.groups[value[0]] = strat
+                strat = hs.shared(strat, key=value[0])
```

Comment:
Actually it's even worse than that; the key is the internal group *index* so named group references are equally unsafe.

Instead of trying to wrap or pass around *another* per-example context object, I've simply made the key a tuple of `(pattern, group_index)` - this seems less error-prone to me.